src/ucp/core/ucp_request.c

```diff
@@ -169,3 +169,96 @@ ucs_status_t ucp_request_start_send(ucp_request_t *req)
     return status;
 }
 
+static UCS_F_ALWAYS_INLINE
+void ucp_iov_buffer_memh_dereg(uct_md_h uct_md, uct_mem_h *memh,
+                               size_t count)
+{
+    size_t it;
+
+    for (it = 0; it < count; ++it) {
+        if (memh[it] != UCT_INVALID_MEM_HANDLE) {
+            uct_md_mem_dereg(uct_md, memh[it]);
+        }
+    }
+}
+
+ucs_status_t ucp_request_send_buffer_reg(ucp_request_t *req, ucp_lane_index_t lane)
+{
+    uct_md_h uct_md         = ucp_ep_md(req->send.ep, lane);
+    ucp_frag_state_t *state = &req->send.state;
+    size_t iov_it, iovcnt;
+    const ucp_dt_iov_t *iov;
+    uct_mem_h *memh;
+    ucs_status_t status;
+
+    status = UCS_OK;
+    switch (req->send.datatype & UCP_DATATYPE_CLASS_MASK) {
+    case UCP_DATATYPE_CONTIG:
+        status = uct_md_mem_reg(uct_md, (void *)req->send.buffer, req->send.length,
```

Comment:
@yosefe The zero memory registration avoided in uct_md_mem_reg() function. In case of `contig` datatype it has no check in UCP (IOV does it at line 209 here).